dbms/src/Interpreters/Join.cpp

```diff
@@ -1460,91 +1523,109 @@ void Join::finishOneBuild()
-        /// checking all blocks.
-        /// So there is a tradeoff between returning quickly and avoiding waste.
-        ///
-        /// If all rows have null key, test results at the time of writing show that the time consumed by checking null rows is
-        /// several times than that of checking all blocks, and it increases as the number of rows in the right table increases.
-        /// For example, the number of rows in right table is 2k and 20000k in left table, null rows take about 1 time as long
-        /// as all blocks. When the number of rows in right table is 5k, 1.4 times. 10k => 1.7 times. 20k => 1.9 times.
-        ///
-        /// Given that many null rows should be a rare case, let's use 2 times to simplify thinking.
-        /// So if null rows occupy 1/3 of all rows, the time consumed by null rows and all blocks are basically the same.
-        /// I choose 1/3 as the cutoff point. If null rows occupy more than 1/3, we should check all blocks directly.
-        if (unlikely(is_test))
-            null_key_check_all_blocks_directly = false;
-        else
-            null_key_check_all_blocks_directly = static_cast<double>(null_rows_size) > static_cast<double>(total_input_build_rows) / 3.0;
-    }
+        finalizeNullAwareSemiFamilyBuild();
 
     if (isCrossJoin(kind))
-    {
-        original_blocks.clear();
-        for (const auto & block : blocks)
-            original_blocks.push_back(block);
-        right_rows_to_be_added_when_matched_for_cross_join = 0;
-        for (const auto & block : original_blocks)
-            right_rows_to_be_added_when_matched_for_cross_join += block.rows();
-        if (strictness == ASTTableJoin::Strictness::Any)
-        {
-            /// for cross any join, at most 1 row is added
-            right_rows_to_be_added_when_matched_for_cross_join = std::min(right_rows_to_be_added_when_matched_for_cross_join, 1);
-        }
-        else if (blocks.size() > 1 && right_rows_to_be_added_when_matched_for_cross_join <= max_block_size)
-        {
-            /// for cross all join, if total right rows is less than max_block_size, then merge all
-            /// the right blocks into one block
-            blocks.clear();
-            auto merged_block = vstackBlocks(std::move(original_blocks));
-            original_blocks.clear();
-            blocks.push_back(merged_block);
-            original_blocks.push_back(merged_block);
-        }
-        /// since shallow_copy_probe_threshold is at least 1, if strictness is any, it will never use SHALLOW_COPY_RIGHT_BLOCK
-        cross_probe_mode = right_rows_to_be_added_when_matched_for_cross_join > shallow_copy_cross_probe_threshold
-            ? CrossProbeMode::SHALLOW_COPY_RIGHT_BLOCK
-            : CrossProbeMode::DEEP_COPY_RIGHT_BLOCK;
-        LOG_DEBUG(log, "Cross join will use {} probe mode", magic_enum::enum_name(cross_probe_mode));
-    }
+        finalizeCrossJoinBuild();
+
+    // set rf is ready
+    finalizeRuntimeFilter();
```

Comment:
Moved here to avoid runtime filter getting stuck with spill

Reply:
I think for now, spill should be disabled explicitly if runtime filter is enabled because runtime filter is not support spill, and for current implementation, for partitions that is spilled, runtime filter is not constructed, which will cause wrong result.

Reply:
ok, updated the code as
```
    if (isEnableSpill())
    {
        if (isSpilled())
        {
            // TODO support runtime filter with spill.
            cancelRuntimeFilter("Currently runtime filter is not compatible with join spill, so cancel runtime filter here.");
        }
        else
        {
            // set rf is ready
            finalizeRuntimeFilter();
        }
        ...
    else
    {
            // set rf is ready
            finalizeRuntimeFilter();
    }
```